dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -543,6 +573,46 @@ private List<Event> inlineLLVMSaturatedAdd(ValueFunctionCall call) {
         );
     }
 
+    private List<Event> inlineLLVMSAddWithOverflow(ValueFunctionCall call) {
+        return inlineLLVMSOpWithOverflow(call, IOpBin.ADD);
+    }
+
+    private List<Event> inlineLLVMSSubWithOverflow(ValueFunctionCall call) {
+        return inlineLLVMSOpWithOverflow(call, IOpBin.SUB);
+    }
+
+    private List<Event> inlineLLVMSMulWithOverflow(ValueFunctionCall call) {
+        return inlineLLVMSOpWithOverflow(call, IOpBin.MUL);
+    }
+
+    private List<Event> inlineLLVMSOpWithOverflow(ValueFunctionCall call, IOpBin op) {
+        final Register resultReg = call.getResultRegister();
+        final List<Expression> arguments = call.getArguments();
+        final Expression x = arguments.get(0);
+        final Expression y = arguments.get(1);
+        assert x.getType() == y.getType();
+
+        final IntegerType iType = (IntegerType) x.getType();
+        final Expression sum = expressions.makeBinary(x, op, y);
+        final Expression flag = expressions.makeCast(
+                expressions.makeNot(checkIfValueInRangeOfType(sum, iType, true)),
+                types.getIntegerType(1)
+        );
+
+        return List.of(
+                EventFactory.newLocal(resultReg, expressions.makeConstruct(List.of(sum, flag)))
+        );
+    }
+
+    private Expression checkIfValueInRangeOfType(Expression value, IntegerType integerType, boolean signed) {
+        final Expression minValue = expressions.makeValue(integerType.getMinimumValue(signed), integerType);
+        final Expression maxValue = expressions.makeValue(integerType.getMaximumValue(signed), integerType);
+        return expressions.makeAnd(
+                expressions.makeLTE(minValue, value, true),
+                expressions.makeLTE(value, maxValue, true)
+        );
+    }
```

Comment:
Now here is the biggest problem of why I didn't wanna merge the overflow checking right away.
This code does only work when doing integer encoding, meaning that with the standard BV encoding we won't detect overflows.
A correct implementation would extend the bit-width, perform the arithmetic, check for overflows, and then truncate.

From LLVM's language manual:
```
An operation overflows if, for any values of its operands A and B and for any N larger than
the operands’ width, ext(A op B) to iN is not equal to (ext(A) to iN) op (ext(B) to iN) where 
ext is sext for signed overflow and zext for unsigned overflow, and op is the 
underlying arithmetic operation.
```

Reply:
I kinda dislike that we still have this weird range check, but I see that we need it.
It's like this: we generate encoding `A & B` and when doing integer encoding, only `A` is relevant and `B` is trivial, but when doing BV encoding `A` is trivial and `B` is relevant. Yet we encode both at all times.
Maybe the SMT solver detects this and removes the trivial part in either case.

Well, we can keep it like this, but I think we should document that this range-check is only relevant when doing integer-based reasoning/encoding and pointless for BVs.